docs/proposals/202003_thanos_rules_federation.md

```diff
@@ -0,0 +1,220 @@
+status: accepted
+owner: s-urbaniak
+---
+
+### Related Tickets
+
+https://github.com/thanos-io/thanos/issues/1375
+
+## Summary
+
+Thanos allows a global query view for the Prometheus series. This means discovering, connecting various (often remote) "leafs" components and aggregating series data from them.
+
+Since we work with metrics, evaluating recording rules and alerts is a very important part of our system. Some of our "StoreAPIs" like Prometheus and Thanos Ruler are designed for that. However, we currently don’t have a way to present those resources in a federated way e.g Alerts and Recording Rules. This document explores the potential way of solving this in the Prometheus Ecosystem.
+
+## Motivation
+
+Deployment scenarios with various levels of federated setups are not unusual. Some deployments leverage Thanos Ruler to query multiple data sources (i.e. via Thanos Querier) for rule and alert evaluation. This topology sacrifices availability. At the same time in the same deployment topology, a dedicated Prometheus instance is deployed for certain critical alerting rules implying high availability. Here, we are missing a federation middleware exposing a consolidated view of evaluated rules and alerts in the underlying Prometheus and Thanos Ruler instances.
+
+Pitfalls of current solutions:
+* It’s tedious to manually visit each leaf and we are lazy.
+* Visiting leaf instances (be it Thanos Ruler or dedicated Prometheus endpoints) have to be implemented manually.
+
+## Goals
+
+* Present other Prometheus/Thanos Resources in a global view.
+    * This also means an up-to-date view (e.g statuses)
+* Simple to run if you already run Thanos.
+    * Use consistent protocols (if we suddenly switch to pure HTTP from gRPC it will break user’s proxies for auth, routing, monitoring, rate limiting, etc)
+    * Potentially use the same connections, discovered targets by Querier.
+
+## Verification
+
+* Unit tests which would fire up a dummy Rules API and check different scenarios.
+* Ad-hoc testing.
+
+## Proposal
+
+### Rules API
+
+We propose to add following proto service and propagate all using existing components that have rules, so:
+* Sidecar
+* Querier (for federation)
+* Ruler
+
+The newly introduced `Rules` service is designed to reflect [Prometheus' Rules API](https://github.com/prometheus/prometheus/blob/bc703b64568ebfaecf27b9b70be737ad318e217a/web/api/v1/api.go) allowing to retrieve recording and alerting rules.
+
+```proto
+service Rules {
+  /// Rules has info for all rules.
+  rpc Rules(RulesRequest) returns (stream RulesResponse);
+}
+```
+
+Sidecar proxies the Rules request to its local Prometheus instance and synthesizes the response. Similarly, Thanos Ruler constructs the response based on its local state.
+
+Thanos Querier fans-out to all know Rules endpoints configured via `--rule` command line flag, merges and deduplicates the result. Generally the deduplication logic is less complex than with time series, specifically:
```

Comment:
There is missing decision item:

Can user specify --rule X without --store X? Is this planned to be implemented?

Reply:
@bwplotka i suggest to have completely independent settings as a future enhancement. For now, we initially discussed to have rule endpoints be a strict subset of store endpoints which makes the initial code handling much easier (piggy packing on the existing store health checks)

Reply:
Concrete implication for the initial implementation would be: if user does specify `--rule` which do not fit as a subset of the given `--store` (and other related store discover settings), we'd log out that fact. Does that sound reasonable/

Reply:
Could be as a start yes, but it's not stated yet clear here. (: Can we clarify?

Reply:
Clarified :+1: :green_heart: